src/sources/docker_logs/mod.rs

```diff
@@ -994,25 +993,38 @@ impl ContainerLogInfo {
         let timestamp_str = splitter.next()?;
         let timestamp = match DateTime::parse_from_rfc3339(timestamp_str) {
             Ok(timestamp) => {
-                // Timestamp check
+                // Timestamp check. This is included to avoid processing the same log multiple times, which can
+                // occur when a container changes generations, and to avoid processing logs with timestamps before
+                // the created timestamp.
                 match self.last_log.as_ref() {
-                    // Received log has not already been processed.
-                    Some(&(ref last, gen))
-                        if *last < timestamp || (*last == timestamp && gen == self.generation) =>
-                    {
-                        // noop
-                    }
-                    // Received log is after the time the container was created.
-                    None if self.created <= timestamp.with_timezone(&Utc) => {
-                        // noop
+                    Some(&(last, gen)) => {
+                        if last < timestamp || (last == timestamp && gen == self.generation) {
+                            // Noop - log received in order.
+                        } else {
+                            // Docker returns logs in order.
+                            // If we reach this state, this log is from a previous generation of the container.
+                            // It was already processed, so we can safely skip it.
+                            trace!(
+                                message = "Received log from previous container generation.",
+                                log_timestamp = %timestamp_str,
+                                last_log_timestamp = %last,
+                            );
+                            return None;
+                        }
                     }
-                    // Received log is older than the previously received entry.
-                    _ => {
-                        emit!(DockerLogsReceivedOutOfOrderError {
-                            container_id: self.id.as_str(),
-                            timestamp_str,
-                        });
-                        return None;
+                    None => {
+                        if self.created < timestamp.with_timezone(&Utc) {
+                            // Noop - first log to process.
+                        } else {
+                            // Received a log with a timestamp before that provided to the Docker API.
+                            // This should not happen, but if it does, we can just ignore these logs.
```

Comment:
I'd argue that if this should not happen - but somehow it does, then we should be raising an error.

Reply:
It *shouldn't* happen from my understanding of the Docker API. However, in the case that this does occur, I would argue that it is not an error that we need to surface to the user.  The Docker Logs source states that it only processes logs with timestamps after the time that it was started. As such, it should ignore these logs by design.